# Controller Manager Parameters
controller_manager:
  ros__parameters:
    update_rate: 100 # Frequency of the controller updates in Hz

    

# Joint State Controller to publish all joint states
husky_joint_publisher:
  type: "joint_state_controller/JointStateController"
  publish_rate: 50





# Differential Drive Controller for wheel control
husky_velocity_controller:
  type: "diff_drive_controller/DiffDriveController"
  left_wheel: ['front_left_wheel_joint', 'rear_left_wheel_joint', 'mid_front_left_wheel_joint', 'mid_rear_left_wheel_joint']
  right_wheel: ['front_right_wheel_joint', 'rear_right_wheel_joint', 'mid_front_right_wheel_joint', 'mid_rear_right_wheel_joint']
  publish_rate: 50
  pose_covariance_diagonal: [0.001, 0.001, 0.001, 0.001, 0.001, 0.03]
  twist_covariance_diagonal: [0.001, 0.001, 0.001, 0.001, 0.001, 0.03]
  cmd_vel_timeout: 0.25
  velocity_rolling_window_size: 2

  # Publish the final output cmd_vel to /husky_velocity_controller/cmd_vel_out
  publish_cmd: true

  # Base frame ID for odometry
  base_frame_id: base_link

  # Odometry fused with IMU is published by robot_localization
  enable_odom_tf: false

  # Wheel separation and radius multipliers
  wheel_separation_multiplier: 1.0 # default: 1.0
  wheel_radius_multiplier: 1.0 # default: 1.0

  # Velocity and acceleration limits for the robot
  linear:
    x:
      has_velocity_limits: true
      max_velocity: 5.0   # m/s
      has_acceleration_limits: true
      max_acceleration: 3.0   # m/s^2
  angular:
    z:
      has_velocity_limits: true
      max_velocity: 2.0   # rad/s
      has_acceleration_limits: true
      max_acceleration: 6.0   # rad/s^2

# Velocity Controllers for Individual Wheels
rear_left_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "rear_left_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0

rear_right_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "rear_right_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0

front_left_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "front_left_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0

front_right_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "front_right_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0
    
    
    
#Middle wheels:

mid_rear_left_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "mid_rear_left_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0

mid_rear_right_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "mid_rear_right_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0

mid_front_left_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "mid_front_left_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0

mid_front_right_wheel_velocity_controller:
  type: "velocity_controllers/JointVelocityController"
  joint: "mid_front_right_wheel_joint"
  pid:
    p: 10.0
    i: 0.0
    d: 0.01
    i_clamp: 0.0
